src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs

```diff
@@ -637,10 +639,21 @@ private void ProcessEvents()
                             AddDirectoryWatch(associatedDirectoryEntry, nextEvent.name);
                         }
 
-                        const Interop.Sys.NotifyEvents switchMask =
-                            Interop.Sys.NotifyEvents.IN_IGNORED |Interop.Sys.NotifyEvents.IN_CREATE | Interop.Sys.NotifyEvents.IN_DELETE |
-                            Interop.Sys.NotifyEvents.IN_ACCESS | Interop.Sys.NotifyEvents.IN_MODIFY | Interop.Sys.NotifyEvents.IN_ATTRIB |
-                            Interop.Sys.NotifyEvents.IN_MOVED_FROM | Interop.Sys.NotifyEvents.IN_MOVED_TO;
+                        // Check if the event should have been filtered but was unable because of inotify's inability
+                        // to filter files vs directories.
+                        const Interop.Sys.NotifyEvents fileDirEvents = Interop.Sys.NotifyEvents.IN_CREATE |
+                                Interop.Sys.NotifyEvents.IN_DELETE |
+                                Interop.Sys.NotifyEvents.IN_MOVED_FROM |
+                                Interop.Sys.NotifyEvents.IN_MOVED_TO;
+                        if ((((uint)fileDirEvents & mask) > 0) &&
```

Comment:
The reason the mask is split is so that the `NotifyEvents` that are unrelated to the toggling of `DirectoryName` or `FileName` do not get excluded based on those values.

For example, if we receive an IN_MODIFY event in which the target path is a file, then we don't want to ignore that event even if `NotifyFilters.FileName & mask == 0`. In the modified case, the only `NotifyFilters` flag that matters are the ones that enable that event: `NotifyFilters.LastAccess`, `NotifyFilters.Size`, etc.

This behavior can be seen in [this test](https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.File.NotifyFilter.cs#L147) where the behavior is the same across platforms with the exception that Unix is less precise in its filtering.